### PR TITLE
Update loom and gradle

### PR DESCRIPTION
**Pull request by modmuss50**
- Upgrading to loom 0.7 and the new run config system
- Update to gradle 7
- Start looking at singing the jars - currently just pgp. jar singing is TODO
- Build reproducable jars
- Build testmod jars - a fat jar depends on some loom work
- Incease default mem to 4G
- Enable parallel by default

### CODE REVIEW

**Review by liach — APPROVED** [review]

**Comment by LambdAurora** [comment]
I thought loom 0.7 wasn't ready and gradle 7 is not even entirely released yet?

**Comment by modmuss50** [comment]
Most of the work that went into 0.7 was around getting a project like fabric API working on gradle 7 and creating the jars for the testmods.

This will work on gradle 6 but by the time this is ready to be merged I expect that gradle 7 will be out, release of loom 0.7 depends on this being tested as well.

**Comment by modmuss50** [comment]
This has now been updated to gradle 7.0 release

File: build.gradle [hunk 1]
```diff
@@ -6,14 +6,16 @@ buildscript {
 
 plugins {
 	id "java"
+	id "java-library"
```
Comment (Juuxel) [inline build.gradle:9]:
```suggestion
	id "java-library"
```
`java-library` applies `java` (and I think Loom does too)

File: build.gradle [hunk 2]
```diff
@@ -6,14 +6,16 @@ buildscript {
...
 	id "eclipse"
 	id "idea"
```
Comment (Juuxel) [inline build.gradle:11]: Loom also applies these two so they're not really needed.

File: build.gradle (hunk 1 above)
Comment (maityyy) [reply build.gradle:9]: Loom is already using `java`, yes.

Comment (i509VCB) [reply build.gradle:9]: Java library is where you'll get the api configuration

**Comment by modmuss50** [comment]
This is currenly broken with

```
Build file '/Users/mark/Documents/fabric-loom/fabric/build.gradle' line: 202

A problem occurred configuring project ':fabric-api-base'.
> Cannot change dependencies of dependency configuration ':api' after it has been included in dependency resolution.
```

The whole thing (down into loom as well) is a tottal mess of afterEvaulates expected to run in a certain order, im not sure how this even really worked before.

**Comment by modmuss50** [comment]
After fixing that, another similar issue:

```
A problem occurred configuring root project 'fabric-api'.
> Could not resolve all dependencies for configuration ':compileClasspath'.
   > A problem occurred configuring project ':fabric-api-base'.
      > Failed to notify project evaluation listener.
         > Task with name 'remapAllSources' not found in root project 'fabric-api'.
         > Task with name 'remapAllTestmodSources' not found in root project 'fabric-api'.
```

**Comment by modmuss50** [comment]
Im not really too sure how this ever worked before, and what to do about fixing it. A propper fix to make this work properly is months of loom work.

Its kinda blocking me being happy with 0.7 loom to call it stable

**Comment by Technici4n** [comment]
Doesn't the problem lie within the fabric api buildscript? I'm not sure because I don't know how Gradle or Loom work.

**Comment by liach** [comment]
Hmm, when was `remapAllSources` task registered? Does its registration depend on a lot of afterEvaluate code?

**Comment by modmuss50** [comment]
Yes, the whole thing (this buildscript and loom) depends on a lot of fragile afterEvaluate code. Its made even harder by trying to enable parrell builds by default.

**Comment by Toffikk** [comment]
I managed to build fabric-api with loom 0.7 and gradle 7.0 , but without signing the jars 
[build.gradle.txt](https://github.com/FabricMC/fabric/files/6370461/build.gradle.txt)

windows works too

**Comment by modmuss50** [comment]
What is here is tempermental, I can always reproduct it on my laptop (8 core MacOS) after deleting the .fabric folder. Subsquent runs will work if the .fabric is there.

File: build.gradle [hunk 3]
```diff
@@ -139,36 +151,62 @@ allprojects {
 	}
 
 	jar {
-		classifier = "dev"
+		archiveClassifier = "dev"
+	}
+
+	artifacts {
+		dev file: file("${project.buildDir}/libs/$archivesBaseName-${version}-dev.jar"), type: "jar", builtBy: jar
```
Comment (Toffikk) [inline build.gradle:158]: Use
${project.version} instead of ${version}

**Review by Toffikk — CHANGES_REQUESTED** [review]

File: build.gradle [hunk 4]
```diff
@@ -139,36 +151,62 @@ allprojects {
...
+
 	afterEvaluate {
 		remapJar {
-			input = file("${project.buildDir}/libs/$archivesBaseName-${version}-dev.jar")
-			archiveName = "${archivesBaseName}-${version}.jar"
-		}
-
-		artifacts {
-			dev file: file("${project.buildDir}/libs/$archivesBaseName-${version}-dev.jar"), type: "jar", builtBy: jar
+			input = file("${project.buildDir}/libs/$archivesBaseName-${archiveVersion.get()}-dev.jar")
```
Comment (Toffikk) [inline build.gradle:183]: Don’t use
${archiveVersion.get()} , use ${project.version}

**Review by Toffikk — CHANGES_REQUESTED** [review]

File: build.gradle [hunk 5]
```diff
@@ -139,36 +151,62 @@ allprojects {
...
+			archiveFileName = "${archivesBaseName}-${archiveVersion.get()}.jar"
```
Comment (Toffikk) [inline build.gradle:184]: The same as above

**Review by Toffikk — CHANGES_REQUESTED** [review]

File: build.gradle [hunk 6]
```diff
@@ -263,16 +312,29 @@ subprojects {
...
+	}
 }
 
-task remapMavenJar(type: net.fabricmc.loom.task.RemapJarTask, dependsOn: jar) {
+task remapMavenJar(type: RemapJarTask, dependsOn: jar) {
 	afterEvaluate {
-		input = file("${project.buildDir}/libs/${archivesBaseName}-${version}-dev.jar")
-		archiveName = "${archivesBaseName}-${version}-maven.jar"
+		input = file("${project.buildDir}/libs/${archivesBaseName}-${archiveVersion.get()}-dev.jar")
+		archiveFileName = "${archivesBaseName}-${archiveVersion.get()}-maven.jar"
```
Comment (Toffikk) [inline build.gradle:328]: The same as above for all cases

**Review by Toffikk — CHANGES_REQUESTED** [review]

File: build.gradle (hunk 3 above)
Comment (liach) [reply build.gradle:158]: I don't see anything in the scope that provides a `version`, and `project` is a default scope offered for `build.gradle` script.

Comment (Juuxel) [reply build.gradle:158]: Note that you can just have `$project.buildDir` and `$version` without the curly braces. (like `$archivesBaseName`)

**Comment by Toffikk** [comment]
Maybe clone this PR Now?

**Comment by Toffikk** [comment]
*close

**Comment by Toffikk** [comment]
Because it’s outdated

**Comment by modmuss50** [comment]
Yes, im going to close this. This PR does have some extra features such as building test mod jars but that can come later.

I based my changes in the snapshot from this PR 👍

**Pull request closed without merging**